File: rust/day14_part2_2019.rs

use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead};
use std::str::FromStr;

#[derive(Debug)]
struct Reaction {
    inputs: Vec<(String, i64)>,
    output: (String, i64),
}

fn parse_reaction(line: &str) -> Result<Reaction, String> {
    let parts: Vec<&str> = line.split(" => ").collect();
    if parts.len() != 2 {
        return Err("Invalid reaction format".to_string());
    }

    let inputs_str = parts[0];
    let output_str = parts[1];

    let inputs: Vec<(String, i64)> = inputs_str
        .split(", ")
        .map(|s| {
            let parts: Vec<&str> = s.split(' ').collect();
            if parts.len() != 2 {
                Err("Invalid input format".to_string())
            } else {
                let amount = i64::from_str(parts[0])
                    .map_err(|_| "Invalid amount".to_string())?;
                Ok((parts[1].to_string(), amount))
            }
        })
        .collect::<Result<_, _>>()?;

    let output_parts: Vec<&str> = output_str.split(' ').collect();
    if output_parts.len() != 2 {
        return Err("Invalid output format".to_string());
    }
    let output_amount = i64::from_str(output_parts[0])
        .map_err(|_| "Invalid output amount".to_string())?;
    let output_name = output_parts[1].to_string();

    Ok(Reaction {
        inputs,
        output: (output_name, output_amount),
    })
}


fn calculate_ore_needed(
    reactions: &HashMap<String, Reaction>,
    fuel_needed: i64,
) -> i64 {
    let mut inventory: HashMap<String, i64> = HashMap::new();
    let mut ore_needed = 0;
    let mut requirements: Vec<(String, i64)> = vec![("FUEL".to_string(), fuel_needed)];

    while let Some((chemical, amount_needed)) = requirements.pop() {
        if chemical == "ORE" {
            ore_needed += amount_needed;
            continue;
        }

        let available = *inventory.get(&chemical).unwrap_or(&0);
        if available >= amount_needed {
            inventory.insert(chemical, available - amount_needed);
            continue;
        }
        let remaining_needed = amount_needed - available;

       
        if let Some(reaction) = reactions.get(&chemical) {
           let multiplier = (remaining_needed + reaction.output.1 - 1) / reaction.output.1;
           
           inventory.insert(chemical, available + multiplier * reaction.output.1 - amount_needed );

            for (input_chem, input_amount) in &reaction.inputs {
                requirements.push((input_chem.clone(), input_amount * multiplier));
            }

        } else {
            panic!("No reaction for chemical: {}", chemical);
        }

    }
    ore_needed
}



fn solve_part2(reactions: &HashMap<String, Reaction>) -> i64 {
    let total_ore: i64 = 1_000_000_000_000;
    let mut low: i64 = 1;
    let mut high: i64 = total_ore;
    let mut max_fuel = 0;

    while low <= high {
        let mid = low + (high - low) / 2;
        let ore_for_mid = calculate_ore_needed(reactions, mid);

        if ore_for_mid <= total_ore {
            max_fuel = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    max_fuel
}


fn main() -> io::Result<()> {
    let file = File::open("input.txt")?;
    let reader = io::BufReader::new(file);

    let mut reactions: HashMap<String, Reaction> = HashMap::new();

    for line_result in reader.lines() {
        let line = line_result?;
        let reaction = parse_reaction(&line).map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
        reactions.insert(reaction.output.0.clone(), reaction);
    }
    

    let ore_needed_for_one_fuel = calculate_ore_needed(&reactions, 1);

    println!("Part 1: Minimum ORE required for 1 FUEL: {}", ore_needed_for_one_fuel);
    println!("Part 2: Max FUEL with 1 trillion ORE: {}", solve_part2(&reactions));

    Ok(())
}

File: csharp/day14_part2_2019.cs

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

public class ChemicalReaction
{
    public Dictionary<string, long> Inputs { get; set; }
    public string OutputName { get; set; }
    public long OutputAmount { get; set; }

    public ChemicalReaction()
    {
        Inputs = new Dictionary<string, long>();
    }
}

public class Program
{
    private static Dictionary<string, ChemicalReaction> reactions = new Dictionary<string, ChemicalReaction>();
    private static Dictionary<string, long> surplus = new Dictionary<string, long>();

    public static void Main(string[] args)
    {
        var lines = File.ReadAllLines("input.txt");
        ParseReactions(lines);

        if (!reactions.ContainsKey("FUEL"))
        {
            Console.WriteLine("Error: FUEL reaction not found in input");
            return;
        }

        var oreAvailable = 1000000000000L;
        var low = 0L;
        var high = oreAvailable;

        // Estimate a reasonable upper bound for high
        surplus.Clear();
        var oreForOneFuel = CalculateOre("FUEL", 1);
        if (oreForOneFuel > 0)
        {
            high = (oreAvailable / oreForOneFuel) * 2;
            if (high == 0) high = 1;
        }

        var maxFuel = 0L;
        while (low <= high)
        {
            var mid = low + (high - low) / 2;
            if (mid == 0)
            {
                low = 1;
                continue;
            }

            surplus.Clear();
            var oreNeeded = CalculateOre("FUEL", mid);

            if (oreNeeded <= oreAvailable)
            {
                maxFuel = mid;
                low = mid + 1;
            }
            else
            {
                high = mid - 1;
            }
        }

        Console.WriteLine(maxFuel);
    }

    private static void ParseReactions(string[] lines)
    {
        foreach (var line in lines)
        {
            var parts = line.Trim().Split(new[] { "=>" }, StringSplitOptions.None);
            var inputs = parts[0].Trim().Split(',');
            var output = parts[1].Trim().Split(' ');

            var reaction = new ChemicalReaction
            {
                OutputName = output[1],
                OutputAmount = long.Parse(output[0])
            };

            foreach (var input in inputs)
            {
                var inputParts = input.Trim().Split(' ');
                reaction.Inputs.Add(inputParts[1], long.Parse(inputParts[0]));
            }

            reactions[reaction.OutputName] = reaction;
        }
    }

    private static long CalculateOre(string chemical, long amount)
    {
        if (chemical == "ORE")
        {
            return amount;
        }

        if (surplus.ContainsKey(chemical) && surplus[chemical] >= amount)
        {
            surplus[chemical] -= amount;
            return 0;
        }

        amount -= surplus.ContainsKey(chemical) ? surplus[chemical] : 0;
        surplus[chemical] = 0;

        if (!reactions.TryGetValue(chemical, out var reaction))
        {
            Console.WriteLine($"Error: Reaction not found for {chemical}");
            return 0;
        }

        var reactionOutputAmount = reaction.OutputAmount;
        var times = (amount + reactionOutputAmount - 1) / reactionOutputAmount;

        long oreNeeded = 0;
        foreach (var input in reaction.Inputs)
        {
            oreNeeded += CalculateOre(input.Key, input.Value * times);
        }

        surplus[chemical] = times * reactionOutputAmount - amount;
        return oreNeeded;
    }
}
